core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -1230,7 +1230,10 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
 
             if (node.getFilter().isPresent()) {
                 Expression expression = node.getFilter().get();
-                process(expression, context);
+                Type type = process(expression, context);
+                if (!type.equals(BOOLEAN)) {
+                    throw semanticException(TYPE_MISMATCH, expression, "Filter expression must evaluate to a boolean: actual type %s", type);
+                }
```

Comment:
This breaks queries with `... FILTER (WHERE <expr>)` where `<expr>` is known to evaluate to `NULL` at analysis time. E.g., 

```
SELECT count(*) FILTER (WHERE NULL) FROM (VALUES 1,2,3) t(x)
```

Take a look at `analyzeWhere` in `SemanticAnalyzer` for an example of how to handle this check. Also, please add a test for that case.

Reply:
@martint NULL literal is not allowed in FILTER clause:
https://github.com/trinodb/trino/blob/67327deef946bb731cd9af4838e92fc5ae27706d/core/trino-main/src/main/java/io/trino/sql/planner/plan/FilterNode.java#L46-L48

Reply:
FilterNode is an IR operation to represent the application of a filter, and is unrelated to the FILTER clause in an aggregation. 

What’s not allowed is for the IR to contain a NullLiteral without an intervening cast that indicates what type it’s supposed to be, since that information is not included in the NullLiteral class.

Reply:
It does work, BTW. What query did you try?

```
trino> SELECT count(*) FILTER (WHERE NULL) FROM (VALUES 1,2,3) t(x);
 _col0
-------
     0
(1 row)
```

Reply:
> is unrelated to the FILTER clause in an aggregation

Ah, I see. It's for flat aggregations. Thanks.

Reply:
Updated the PR.